pkg/platform/api/graphql/model/buildplanner/buildExpression.go

```diff
@@ -151,51 +190,101 @@ func (bx *BuildExpression) Update(operation Operation, requirement Requirement)
-	for i, req := range bx.requirementsNode {
-		r := req.(map[string]interface{})
-		if r[RequirementNameKey] == requirement.Name && r[RequirementNamespaceKey] == requirement.Namespace {
-			bx.requirementsNode = append(bx.requirementsNode[:i], bx.requirementsNode[i+1:]...)
-			(*bx.solveNode)[RequirementsKey] = bx.requirementsNode
+// removeRequirement removes a requirement from the BuildExpression.
+func (bx *BuildExpression) removeRequirement(requirement Requirement) error {
+	for i, req := range bx.requirements {
+		if req.Name == requirement.Name && req.Namespace == requirement.Namespace {
+			bx.requirements = append(bx.requirements[:i], bx.requirements[i+1:]...)
+			(*bx.solveNode)[RequirementsKey] = bx.requirements
 			return nil
 		}
 	}
 
 	return errs.New("Could not find requirement")
 }
 
-func (bx BuildExpression) UpdateRequirement(requirement Requirement) error {
-	for i, req := range bx.requirementsNode {
-		r := req.(map[string]interface{})
-		if r[RequirementNameKey] == requirement.Name && r[RequirementNamespaceKey] == requirement.Namespace {
-			bx.requirementsNode[i] = requirement
-			(*bx.solveNode)[RequirementsKey] = bx.requirementsNode
+// updateRequirement updates an existing requirement in the BuildExpression.
+func (bx BuildExpression) updateRequirement(requirement Requirement) error {
+	for i, req := range bx.requirements {
+		if req.Name == requirement.Name && req.Namespace == requirement.Namespace {
+			bx.requirements[i] = requirement
+			(*bx.solveNode)[RequirementsKey] = bx.requirements
 			return nil
 		}
 	}
 
 	return errs.New("Could not find requirement")
 }
 
+// UpdateTimestamp fetches the latest platform timestamp and updates it in the BuildExpression.
 func (bx BuildExpression) UpdateTimestamp() error {
-	(*bx.solveNode)[AtTimeKey] = time.Now().UTC().Format(time.RFC3339)
+	latest, err := fetchLatestTimeStamp()
+	if err != nil {
+		return errs.Wrap(err, "Could not fetch latest timestamp")
+	}
+
+	formatted, err := time.Parse(time.RFC3339, latest.String())
+	if err != nil {
+		return errs.Wrap(err, "Could not parse latest timestamp")
+	}
+
+	(*bx.solveNode)[AtTimeKey] = formatted
 	return nil
 }
 
-func getRequirementsNode(bx map[string]interface{}) ([]interface{}, error) {
-	solveNode, err := getSolveNode(bx)
+func (bx BuildExpression) MarshalJSON() ([]byte, error) {
+	return json.Marshal(bx.expression)
+}
+
+// fetchLatestTimeStamp fetches the latest timestamp from the inventory service.
+// This function lives in this package to avoid an import cycle.
+func fetchLatestTimeStamp() (*strfmt.DateTime, error) {
+	if condition.InTest() {
+		return &strfmt.DateTime{}, nil
+	}
+
+	client := inventory.Get()
+	result, err := client.GetLatestTimestamp(inventory_operations.NewGetLatestTimestampParams())
 	if err != nil {
-		return nil, errs.Wrap(err, "Could not get solve node")
+		return nil, errs.Wrap(err, "GetLatestTimestamp failed")
 	}
 
+	return result.Payload.Timestamp, nil
+}
```

Comment:
I agree it shouldn't be there, but per the comment, it's there to avoid an import cycle. This stems from us having a graphQL model package and a platform model package. I think addressing this should be part of [this](https://activestatef.atlassian.net/jira/software/c/projects/DX/issues/DX-1899?jql=project%20%3D%20%22DX%22%20AND%20text%20~%20%22model%22%20ORDER%20BY%20created%20DESC) story as it will require a fair amount of reshuffling.

Reply:
I see. In that case, why not feed in the timestamp here?

https://github.com/ActiveState/cli/pull/2559/files#diff-733c58be1cfc77e97f562b5cf09693ccecbf956fd0f3accf8cd9c3ab40d59b09R153

Reply:
Good call!